### PR TITLE
New crafts for build, scan and perm tools

### PR DESCRIPTION
**Pull request by MineShadow7**
# Changes proposed in this pull request:
- Added new craft for build tool
- Added new craft for scan tool
- Added new craft for perm tool

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/ColonyView.java [hunk 1]
```diff
@@ -75,6 +78,11 @@
     private double overallHappiness = 5;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/colony/ColonyView.java:78]: MAJOR Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java [hunk 2]
```diff
@@ -95,10 +98,10 @@ public void toBytes(@NotNull final ByteBuf buf)
      */
     public static class Permission extends AbstractMessage<Permission, IMessage>
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/network/messages/PermissionsMessage.java:99]: MAJOR Reduce this class from 63 to the maximum allowed 30 or externalize it in a public class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2972)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 3]
```diff
@@ -437,6 +443,35 @@ public WindowTownHall(final BuildingTownHall.View townHall)
...
+    }
+
+    /**
+     * Clears and resets all citizens.
+     */
+    private void updateWorkOrders()
+    {
+        workOrders.clear();
+        workOrders.addAll(townHall.getColony().getWorkOrders());
+        workOrders.sort((first, second) -> second.getPriority() > first.getPriority() ? 1 : (second.getPriority() < first.getPriority() ? -1 : 0));
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:472]: MAJOR Extract this nested ternary operation into an independent statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3358)

File: src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java [hunk 4]
```diff
@@ -800,7 +806,7 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
                 }
 
                 rowPane.findPaneOfTypeByID("name", Label.class).setLabelText(name);
-                final boolean isTriggered = townHall.getColony().getPermissions().hasPermission(Permissions.Rank.valueOf(actionsList.getParent().getID().toUpperCase()), action);
+                final boolean isTriggered = townHall.getColony().getPermissions().hasPermission(Rank.valueOf(actionsList.getParent().getID().toUpperCase()), action);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:809]: MINOR Define the locale to be used in this String operation. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1449)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityColonyBuilding.java [hunk 5]
```diff
@@ -290,7 +290,7 @@ public boolean isUseableByPlayer(@NotNull final EntityPlayer player)
     public boolean hasAccessPermission(final EntityPlayer player)
     {
         //TODO This is called every tick the GUI is open. Is that bad?
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityColonyBuilding.java:292]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java [hunk 6]
```diff
@@ -189,13 +185,14 @@ public boolean checkInWareHouse(@NotNull final AbstractBuilding buildingEntry, b
...
                 {
                     buildingEntry.setOnGoingDelivery(true);
-                    list.add(buildingEntry);
+                    taskQueue.add(buildingEntry);
                 }
                 return true;
             }
-            if (list.contains(buildingEntry))
+
+            if (taskQueue.contains(buildingEntry))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:193]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java [hunk 7]
```diff
@@ -92,14 +95,10 @@ public void update()
                 if (i == index)
                 {
                     if(buildingEntry.getValue() instanceof AbstractBuildingWorker
-                            && !list.contains(buildingEntry.getValue())
-                            && ((AbstractBuildingWorker) buildingEntry.getValue()).needsAnything())
-                    {
-                        checkInWareHouse((AbstractBuildingWorker) buildingEntry.getValue(), true);
-                    }
-                    else if(buildingEntry.getValue() instanceof BuildingHome && ((BuildingHome) buildingEntry.getValue()).isFoodNeeded())
+                            && !taskQueue.contains(buildingEntry.getValue())
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:98]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java [hunk 8]
```diff
@@ -117,32 +116,28 @@ public boolean checkInWareHouseForFood(final BuildingHome buildingEntry, final b
...
                 {
                     buildingEntry.setOnGoingDelivery(true);
-                    list.add(buildingEntry);
+                    taskQueue.add(buildingEntry);
                 }
                 return true;
             }
 
-            if (list.contains(buildingEntry))
+            if (taskQueue.contains(buildingEntry))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:124]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

File: src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java [hunk 9]
```diff
@@ -169,14 +164,15 @@ public boolean checkInWareHouse(@NotNull final AbstractBuilding buildingEntry, b
...
                         buildingEntry.setOnGoingDelivery(true);
-                        list.add(buildingEntry);
+                        taskQueue.add(buildingEntry);
                     }
                     return true;
                 }
             }
-            if (list.contains(buildingEntry))
+
+            if (taskQueue.contains(buildingEntry))
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/coremod/tileentities/TileEntityWareHouse.java:173]: MINOR This call to "contains()" may be a performance hot spot if the collection is large. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2250)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 166 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 96 major
* ![MINOR][MINOR] 53 minor
* ![INFO][INFO] 16 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [InventoryCitizen.java#L347](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/inventory/InventoryCitizen.java#L347): Reduce the number of conditional operators (7) used in the expression (maximum allowed 6). [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1067)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L658](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L658): Extract this nested ternary operation into an independent statement. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3358)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L788](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L788): Reduce this anonymous class number of lines from 24 to at most 20, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L885](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L885): Reduce this anonymous class number of lines from 40 to at most 20, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [Colony.java](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java): This file has 1.261 lines, which is greater than 1.000 authorized. Split it into smaller files. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS00104)
1. ![MAJOR][MAJOR] [Colony.java#L44](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L44): Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 43 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
1. ![MAJOR][MAJOR] [Colony.java#L44](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L44): class "Colony" has 84 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [Colony.java#L660](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L660): Reduce the number of returns of this method 11, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
1. ![MAJOR][MAJOR] [Colony.java#L847](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L847): Refactor this method to reduce its Cognitive Complexity from 22 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [Colony.java#L1196](https://github.com/Minecolonies/minecolonies/blob/9a5005b170a8ef837e76325c13ee39c3509e020f/src/main/java/com/minecolonies/coremod/colony/Colony.java#L1196): Assign this magic number 20 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by xavierh** [comment]
Look at the changes, your pr does not look correct

**Comment by marvin-bitterlich** [comment]
wrong direction

**Pull request closed without merging**